### ISSUE TITLE
v1.4 fails when GCR not configured

### ISSUE THREAD

**Issue by sophaskins** [upmc-enterprises/registry-creds#30]
Hiya! I just attempted to set up version 1.4 on one of my clusters - I ended up seeing errors like:

```$ kubectl logs -n kube-system registry-creds-1029257478-hjm48 -f
2017/01/24 15:56:27 Starting up...
2017/01/24 15:56:27 Using AWS Account: XXXXXXXXXXX
2017/01/24 15:56:27 Using AWS Region: us-east-1
2017/01/24 15:56:27 Refresh Interval (minutes): 2
2017/01/24 15:58:27 Refreshing credentials...
2017/01/24 15:58:27 Failed to load ecr credentials: google: could not find default credentials. See https://developers.google.com/accounts/docs/application-default-credentials for more information.
```

the `google: could not find default credentials. See https://developers.google.com/accounts/docs/application-default-credentials for more information.` part of which being puzzling, because I'm not attempting to use GCE at all.

Looking through the code, `registry-creds` iterates through trying both GCR and ECR on `process()` (https://github.com/upmc-enterprises/registry-creds/blob/master/main.go#L205-L216) but if it gets any error from either, it returns that error: https://github.com/upmc-enterprises/registry-creds/blob/master/main.go#L205-L216

I suspect that's the issue - that while I've provided an AWS account id (and thus don't hit this branch: https://github.com/upmc-enterprises/registry-creds/blob/master/main.go#L288-L290), there isn't any actual flagging off of the method not in use.

I *suspect* this also affects the GCR-only path, though it probably manages to set the secrets properly before exiting because it can't reach AWS

I'd be glad to put together some sort of PR for conditionally setting up `secretGenerators` if you like, or if you have a plan already, that'd be rad too. For the mean time, I've deployed pre-rename awsecr-creds 1.2 and things seem to be working fine.

**Comment by stevesloka** [comment]
Hey @sophaskins, super sorry you've hit this. I'm happy to take a look at it in a bit, but if you have time and want to send in a PR, that's 100% cool with me. We can work it out together. =)

**Comment by stevesloka** [comment]
I re-tested everything and it seems to me that `upmcenterprises/registry-creds:1.4` is working fine. Can you post up your replicationController which is backing this? Curious if there's an issue with Env variables vs secrets for data. FWIW, I'm using secrets for all my values.

**Comment by sophaskins** [comment]
I was using a deployment instead of a replicationController - this is what I had:
```
---
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: registry-creds
  namespace: kube-system
  labels:
    version: v1.4
    app: registry-creds
    kubernetes.io/cluster-service: "true"
spec:
  strategy:
    type: Recreate
  selector:
    matchLabels:
      app: registry-creds
  template:
    metadata:
      labels:
        app: registry-creds
    spec:
      containers:
      - image: upmcenterprises/registry-creds:1.4
        name: registry-creds
        imagePullPolicy: Always
        env:
          - name: AWS_ACCESS_KEY_ID
            valueFrom:
              secretKeyRef:
                name: registry-creds
                key: AWS_ACCESS_KEY_ID
          - name: AWS_SECRET_ACCESS_KEY
            valueFrom:
              secretKeyRef:
                name: registry-creds
                key: AWS_SECRET_ACCESS_KEY
          - name: awsaccount
            valueFrom:
              secretKeyRef:
                name: registry-creds
                key: aws-account
```

with the expected values set in a secret. I verified the right values were in the environment by (hackily) just looking at the env of the running container. I just posted the above PR as a sketch of both a failing test demonstrating the behavior, and of the beginning of a possible solution.

I think the root of the problem is that https://github.com/upmc-enterprises/registry-creds/blob/master/main.go#L128-L150 returns an `err` on our system (I'm fairly sure from the first branch). We *don't* have "application default credentials" (https://developers.google.com/identity/protocols/application-default-credentials) because we don't use GCE for anything.

**Comment by sophaskins** [comment]
Resolved in 1.5!
